src/main/java/baritone/cache/FasterWorldScanner.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+
+
+    private List<BlockPos> collectChunkSections(BlockOptionalMetaLookup lookup, Chunk chunk, long chunkX, long chunkZ, int playerSection) {
+        // iterate over sections relative to player
+        List<BlockPos> blocks = new ArrayList<>();
+        ExtendedBlockStorage[] sections = chunk.getBlockStorageArray();
+        int l = sections.length;
+        int i = playerSection - 1;
+        int j = playerSection;
+        for (; i >= 0 || j < l; ++j, --i) {
+            if (j < l) {
+                visitSection(lookup, sections[j], blocks, chunkX, chunkZ);
+            }
+            if (i >= 0) {
+                visitSection(lookup, sections[i], blocks, chunkX, chunkZ);
+            }
+        }
+        return blocks;
+    }
+
+    private void visitSection(BlockOptionalMetaLookup lookup, ExtendedBlockStorage section, List<BlockPos> blocks, long chunkX, long chunkZ) {
+        if (section == null || section.isEmpty()) {
+            return;
+        }
+
+        BlockStateContainer sectionContainer = section.getData();
+        //this won't work if the PaletteStorage is of the type EmptyPaletteStorage
+        if (((IBlockStateContainer) sectionContainer).getStorage() == null) {
+            return;
+        }
+
+        boolean[] isInFilter = getIncludedFilterIndices(lookup, ((IBlockStateContainer) sectionContainer).getPalette());
+        if (isInFilter.length == 0) {
+            return;
+        }
+
+
+        int yOffset = section.getYLocation();
+        BitArray array = ((IBlockStateContainer) section.getData()).getStorage();
+        for (int place : collectBlockLocations(array, isInFilter)) {
+            blocks.add(new BlockPos(
+                chunkX + ((place & 255) & 15),
+                yOffset + (place >> 8),
+                chunkZ + ((place & 255) >> 4)
+            ));
+        }
+    }
+
+    private boolean[] getIncludedFilterIndices(BlockOptionalMetaLookup lookup, IBlockStatePalette palette) {
+        boolean commonBlockFound = false;
+        ObjectIntIdentityMap<IBlockState> paletteMap = getPalette(palette);
+        int size = paletteMap.size();
+
+        boolean[] isInFilter = new boolean[size];
+
+        for (int i = 0; i < size; i++) {
+            IBlockState state = paletteMap.getByValue(i);
+            if (lookup.has(state)) {
+                isInFilter[i] = true;
+                commonBlockFound = true;
+            } else {
+                isInFilter[i] = false;
+            }
+        }
+
+        if (!commonBlockFound) {
+            return new boolean[0];
+        }
+        return isInFilter;
+    }
+
+    private static IntList collectBlockLocations(BitArray array, boolean[] isInFilter) {
+        long[] longArray = array.getBackingLongArray();
+        int arraySize = array.size();
+        int bitsPerEntry = ((IBitArray) array).getBitsPerEntry();
+        long maxEntryValue = ((IBitArray) array).getMaxEntryValue();
+
+        IntList positions = new IntArrayList();
```

Comment:
I think this still has the same problem as https://github.com/cabaletta/baritone/pull/3768#discussion_r1070525845 which is that allocating a new object per every `ExtendedBlockStorage` in render distance could cause a lot of GC pressure?

Reply:
im not really sure this is worth worrying about, maybe im being silly

Reply:
but i do like this approach a lot better than the lambdas! so i think this is definitely worth it (i just didnt like the lambdas to begin with lol)

Reply:
there was an impure per matching block before, so at least now it's only one per chunk section

Reply:
I inlined it anyway so...